### ISSUE TITLE
Gradient of Monte Carlo Expectation using Score Function Estimator

### ISSUE THREAD

**Issue by dionman** [tensorflow/probability#608]
I try to compute the gradient of a Monte Carlo Estimator (corresponding to an ELBO) of the form \grad_{\lambda}ELBO(\lambda) = \grad_{\lambda}\E_{q}[log p(\theta, x) - log q(\theta)]. I express the above using a ```tf.reduce_mean``` over samples of  \lambda. This expression seems to be differentiable in tensorflow and I can get back partial derivatives for \lambda components. Analytically, the gradient can be approximately computed by applying the score function gradient estimator as follows:
 \grad_{\lambda}ELBO \apprx 1/S \sum_{s} [log p(\theta^{(s)}, x) - log q(\theta^{(s)})] \grad_{\lambda} \log q(\theta^{(s)}). 

Is this implicitly done with 
``` gradient(elbo, \lambda) ``` or should I express ELBO differently, e.g. ```tfp.monte_carlo.expectation```. I'm still debugging this, however gradient computation seems to be incorrect at the moment.

**Comment by dionman** [comment]
My question is how we should express an approximations to an expectation which cannot be computed in closed form in tensorflow (so, typically a Monte Carlo average) so that computed gradient is the correct gradient of the expectation (e.g. one that can be expressed as the corresponding MC average after applying the log-derivative trick).
http://blog.shakirm.com/2015/11/machine-learning-trick-of-the-day-5-log-derivative-trick/
